### DISCUSSION TITLE
Questions about MeshGenerator and UserObject Interaction in Actions

### DISCUSSION THREAD

**Discussion by hityyds** [idaholab/moose#26690] [Q&A General] [answered]
Hello everyone,

I recently wrote a MeshGenerator that stores some necessary information in the form of a vector. One of my UserObjects requires this information to perform specific operations. However, I cannot declare a vector variable in the input.i file to receive this information. Therefore, I am considering using actions to add both the MeshGenerator and UserObject and facilitate the exchange of information between them. My knowledge of actions is limited, and I have only attempted to use actions to add kernels and variables in the past. Now, I have three questions:

**1. Immediate Execution of Appended MeshGenerator:**

When using `_app.appendMeshGenerator` in the `act()` function, will the relevant MeshGenerator operations be executed immediately? Or does this function merely declare the MeshGenerator, postponing execution? I initially planned to call `_app.appendMeshGenerator` and then declare a vector object in the action to receive the needed information from the MeshGenerator, passing it to my UserObject. If the MeshGenerator is not executed immediately, how can I obtain the required vector information?

**2. Action Destruction after Task Completion:**

After an action completes all its tasks, is the action automatically destroyed?

**3.  MeshGenerator Lifetime when Declared in input.i:**

If I choose to declare the MeshGenerator in the input.i file, will the MeshGenerator object be destroyed once the mesh generation is completed? This concern arises as I need to access the information saved by the MeshGenerator in the action, necessitating the continued existence of the MeshGenerator object when the action is executed.

Thank you in advance for your guidance and assistance.

**Comment by GiudGiud** [comment, marked as answer]
Hello

First, if you don't mind, let's make sure you actually need to do it this way. 
What is the size of this vector? Is this something that has a value for every element? (if so, a field variable makes more sense. If it s a vector of indices, extra element integer ids, stored on elements, also make more sense). 

How big is this vector? Could it be defined in the input instead?

Does this vector depend on the output of the mesh generation process? Or does it guide the mesh generation process as well as the UO?

1) No appending mesh generators will only schedule it for execution when mesh generation happens. Calling meshgenerator::execute() executes the mesh generator but you should not do that outside of the mesh generation process. 

The mesh generators are actually constructed and executed very early in the simulation. I think an Action would be a fine place to store the vector, as Actions are constructed early as well (unlike user objects)

2) no the actions stick around. We have talked about destructing them but never did

3) Same for mesh generators. We have talked about releasing that memory but have not done it

Guillaume

**Reply by hityyds** [reply]
Yes, I do need to store the data in the form of a vector. 

I'm working with two nested meshes, and the CoarseMeshExtraElementIDGenerator can only map between elements using extra_id. Since I also need to understand the adjacency relationships between different elements in the coarse mesh, I'm attempting to store this information in the form of a `vector<vector<>>` (where vector[coarse_elem_id][coarse_elem_side]=neighbor). 

Previously, I considered saving the coarse mesh (thanks to your previous help, I learned that I can save the coarse mesh as well) and directly providing it to my UserObject. However, it seems that functions directly manipulating the mesh are member functions of MeshGenerator. Implementing this operation in the UserObject might require substantial modifications.

Currently, I have an urgent need, so I opted to use a vector to store this information, which is the most familiar and quickest approach for me. I modified the code of CoarseMeshExtraElementIDGenerator to help me extract this information from the coarse mesh. Therefore, this vector variable depends on the output of the mesh generation but has no influence to its execution.

**Reply by GiudGiud** [reply]
Using the save_in mesh capability, you could retrieve a pointer to elements (from the id) in the coarse mesh and use the ->side_ptr(coarse_side_index) to get the side element. 

Why does the user object have to modify the mesh? The coarse mesh cannot be fully specified at the time of its generation?

I think giving either a pointer to or a copy of this vector of vector to the action is reasonable. Then the UO can retrieve it from there

**Reply by hityyds** [reply]
> Why does the user object have to modify the mesh? The coarse mesh cannot be fully specified at the time of its generation?

This is because, when searching for the definition of the `getMeshByName` function in VSCode, I found that it is a member function of MeshGenerator. It's possible that my search was not thorough enough.

> I think giving either a pointer to or a copy of this vector of vector the action is reasonable. Then the UO can retrieve it from there

Thank you so much for your help.

**Reply by GiudGiud** [reply]
`getMeshByName` is only for mesh generators. 
```

  /**
   * Get the saved mesh by name
   */
  [[nodiscard]] std::unique_ptr<MeshBase> getSavedMesh(const std::string & name);
```
from the MeshGeneratorSystem, should be accessible from anywhere.
You can get the system with 
```
_app.getMeshGeneratorSystem()
```

**Reply by hityyds** [reply]
Thank you once more; this is precisely what I was looking for.
